### ISSUE TITLE
Invoke save callback even if there's nothing to save ...

### ISSUE THREAD

**Issue by apendua** [jagi/meteor-astronomy#101]
... because sometimes I don't have enough data to tell if I am really changing the model or not. But I still want to have a consistent behavior on:
```javascript
// maybe I am not changing anything, I don't know ...
model.set('someField', 'someValue');
model.save(function () {
   // here, I want to be sure that my data is properly saved ...
});
```
@jagi Do you think this behavior makes sense?

**Comment by andrerfneves** [comment]
I believe that's the idea behind the Validators module. If you run a .validate() call on the Astro object, it will follow your validations set up on the schema and it will only save if it validates.

Sort of like this:

Method in the server
```javascript
Meteor.methods({
    'newProject': function(project) {
        if (project.validate()) {
            project.set('author', Meteor.userId());
            project.save(); // Save document onto collection
            return project; // Returns object
        }
    }
});
```
and make these calls on the client events with the .validate call on the object.
```javascript
  'click [name=save]': function(e, tmpl) {
    var user = this;

    if (project.validate()) {
      Meteor.call('/project/save', project, function(err) {
        if (!err) {
          Router.go('projects');
        } else {
          user.catchValidationException(err);
        }
      });
    }
  }
```
And make sure you set up some sort of validation on your schema. Like this:
```javascript
Project = Astro.Class({
  name: 'Project',
  collection: Projects,
  fields: {
    firstName: 'string',
    lastName: 'string',
    email: 'string',
    age: 'number'
  },
  validators: {
    firstName: [
      Validators.required(),
      Validators.string(),
      Validators.minLength(3)
    ],
    lastName: [
      Validators.required(),
      Validators.string(),
      Validators.minLength(3)
    ],
    email: [
      Validators.required(),
      Validators.string(),
      Validators.email(3),
      Validators.unique(),
    ],
    age: [
      Validators.required(),
      Validators.number(),
      Validators.gte(18),
      Validators.lte(100)
    ]
  }
});
```
Use this as reference too https://github.com/jagi/meteor-astronomy-examples

**Comment by jagi** [comment]
When calling:

```js
doc.save();
```

It will only save document if there were any changes. It won't do that if not needed. You don't have to think about it. Just set values of fields and save it. It will save it if needed. If don't, nothing will happen. Does is answer your question?

**Comment by apendua** [comment]
@jagi I know what you're saying, but the problem lies somewhere else.

The following code is a pretty abstract example but it should explain my point:
```javascript
// assume "model" is defined in the scope here
function autosave(fieldName, fieldValue, cb) {
  model.set(fieldName, fieldValue);
  model.save(cb);
}
// ...
// now, I am calling my asynchronous function somewhere else, e.g.
$('.saving.in.progres').show();
autosave('name', 'Tomasz', function () {
   // here, if I am unlucky and I don't really change the name
   // this is never called and my UI will be broken
   $('.saving.in.progress').hide();
});
```
And now more formally. If you expose an asynchronous API by allowing users to pass callbacks to the `save` method they will expect that the callback is eventually called no matter what the current state of the model is. Maybe it should be called immediately when there are no changes to save?

I think that not calling it is a BUG in the API design because it forces the user, who may be working in a totally different layer of code, to alter the way `save` method is called - with or without callback - based on the fact if there were any real changes to the model.

**Comment by jagi** [comment]
Oh I see, you are talking about calling callback function even if there was no need for saving a document. I will take a look how it's done in Meteor and I will do it the same way. Will let you know.

**Comment by jagi** [comment]
Yes, in deed Meteor calls callback function even if there was no save, so I will update it. Thanks for noticing it :)

**Comment by apendua** [comment]
:+1:

**Comment by jagi** [comment]
Bug fixed, it will appear in the 1.0.0-rc.5 which I will probably publish today or tomorrow
